server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -162,34 +184,75 @@ public CachingClusteredClient(
     return new SpecificQueryRunnable<>(queryPlus, responseContext).run(timelineConverter);
   }
 
-  @Override
-  public <T> QueryRunner<T> getQueryRunnerForSegments(final Query<T> query, final Iterable<SegmentDescriptor> specs)
+  private <T> QueryRunner<T> runAndMergeWithTimelineChange(
+      final Query<T> query,
+      final UnaryOperator<TimelineLookup<String, ServerSelector>> timelineConverter
+  )
   {
-    return new QueryRunner<T>()
-    {
-      @Override
-      public Sequence<T> run(final QueryPlus<T> queryPlus, final Map<String, Object> responseContext)
-      {
-        return CachingClusteredClient.this.run(
+    final OptionalLong mergeBatch = QueryContexts.getIntermediateMergeBatchThreshold(query);
+
+    if (mergeBatch.isPresent()) {
+      final QueryRunnerFactory<T, Query<T>> queryRunnerFactory = conglomerate.findFactory(query);
+      final QueryToolChest<T, Query<T>> toolChest = queryRunnerFactory.getToolchest();
+      return (queryPlus, responseContext) -> {
+        final Stream<? extends Sequence<T>> sequences = run(
             queryPlus,
             responseContext,
-            timeline -> {
-              final VersionedIntervalTimeline<String, ServerSelector> timeline2 =
-                  new VersionedIntervalTimeline<>(Ordering.natural());
-              for (SegmentDescriptor spec : specs) {
-                final PartitionHolder<ServerSelector> entry = timeline.findEntry(spec.getInterval(), spec.getVersion());
-                if (entry != null) {
-                  final PartitionChunk<ServerSelector> chunk = entry.getChunk(spec.getPartitionNumber());
-                  if (chunk != null) {
-                    timeline2.add(spec.getInterval(), spec.getVersion(), chunk);
-                  }
-                }
-              }
-              return timeline2;
-            }
+            timelineConverter
+        );
+        return MergeWorkTask.parallelMerge(
+            sequences.parallel(),
+            sequenceStream ->
+                new FluentQueryRunnerBuilder<>(toolChest)
+                    .create(
+                        queryRunnerFactory.mergeRunners(
```

Comment:
This requires to use `ForkJoinPool.managedBlock()` in `SegmentMetadataQueryRunnerFactory.mergeRunners()`, `GroupByMergedQueryRunner.run()`, and `GroupByMergingQueryRunnerV2.run()`.

It would be great if there is a util method to create `ForkJoinPool.managedBlock()` for futures to make the code clearer.

Reply:
Also, I've just noticed that this also breaks the assumption of groupBy v2. In groupBy v2, the broker assumes that the intermediate aggregates are always sorted by the grouping keys, so that it can perform the merge-sorted aggregation. However, calling `QueryRunnerFactory.mergeRunners()` internally performs hash-aggregation (or array-based aggregation) and then sort again which is inefficient. For groupBy v2, merge-sorted aggregation should be performed in parallel. Maybe we need to add a new method to QueryToolChest which is different from the merge in historicals and the final merge in brokers.

We've recently had a discussion about this on dev mailing. See https://lists.apache.org/thread.html/b4c1cbe0c97e52ae5a137f4315af6a202a24d3034f53ce92c0d30150@%3Cdev.druid.apache.org%3E for more details.

Reply:
hmmm, sounds like GroupBy needs some more love to potentially support this feature. @jihoonson since it is an opt-in feature in the query context, would you be ok if it is merged without full group-by support, but with documentation that it probably doesn't work for groupBy queries?

Reply:
> hmmm, sounds like GroupBy needs some more love to potentially support this feature. 

@drcrallen It sounds more to me like the groupBy issues are a symptom of using the wrong API here: the new CachingClusteredClient uses `mergeRunners(ExecutorService, Iterable<QueryRunner<T>>)` when it probably shouldn't. The javadocs for that method say that it should be called on the runners from `createRunner(Segment)`:

```
   * Runners generated with createRunner() and combined into an Iterable in (time,shardId) order are passed
   * along to this method with an ExecutorService.  The method should then return a QueryRunner that, when
   * asked, will use the ExecutorService to run the base QueryRunners in some fashion.
```

There's no guarantee in general that this method will do the right thing if applied to higher level runners.

I could suggest a simple fix that will make things work for groupBy (or any other query type): replace the mergeRunners() call with an explicit creation of ChainedExecutionQueryRunner that uses the mergeFjp. This is already what mergeRunners does for timeseries and topN, so they'll work the same way. For query types that do something more 'special' for mergeRunners, this should still work: it will basically materialize results for each server, then merge them using mergeResults (which knows how to merge those by-server results).

I also have a couple of general questions. If I read it right, this patch is materializing results for each server inside a mergeFjp thread before starting to merge them. That has a couple of possible ill effects:

1. Increased memory usage, especially for queries that might generate a lot of results per server (i.e. scan, groupBy, topN with high threshold, anything that's by segment).
2. Starvation of mergeFjp threads if a few queries take too long to materialize results from their servers, and their materialization tasks are sitting around blocking inside the FJP. Before this patch, a small number of queries could not starve out any of the broker thread pools (although, a large number of slow queries could starve out the broker's http serving thread pool).

I'm wondering if it makes more sense to keep using a single thread to retrieve and deserialize results from historicals, but then merge them in a multithreaded pool, such that we know the merging work won't block? Or, is the deserialization of retrieved results one of the bottlenecks we're trying to avoid? In that case, what about multithreading deserialization too, but still having the retrieval be single-threaded?

Could you talk more about your experience running this in prod, & if you've noticed anything like these effects?

Reply:
I'll comment more later but as a brief addressing of the last points:

There wasn't as much of an effect on the memory usage as I expected once it got into the wild here. I think this is partly due to the fact that the merges get threaded in as soon as they are ready from the nodes, and if your skew on query times across your nodes is comparable to the parallel merging ability, then there isn't much accumulated on the brokers. So if a broker can merge 100 nodes in 0.5 seconds for example, and the variance in node response time on your historicals swings between 0.25s and 0.75s, then by the time the 0.25s ones have been merged in, the 0.75s ones start to stack up to be merged. I don't have anything to prove this hypothesis.

Starvation of the mergeFJP is a very real concern. That is why the `ForkJoinPool.managedBlock` is needed. In such a scenario the FJP detects that too many threads are blocked and adjusts its thread pool appropriately. I don't have strong examples of group-by queries on brokers using this approach, so I haven't exercised that code path yet, but the topN code path seems to be working as expected with this approach without encountering thread starvation.

I'll definitely check out the `ChainedExecutionQueryRunner` approach

Reply:
In addition to @gianm's comment, I wonder the same goal can be achieved by using the processing threads of brokers instead of adding a new ForkJoinPool. The broker already has its processing thread pool, but it's not being used. I think this might be better because we can avoid adding a new configuration for ForkJoinPool which looks similar to that for the existing processing thread pool.